src/diagonal.works/b6/ui/lines.go

```diff
@@ -0,0 +1,707 @@
+}
+
+func fillSubstacksFromError(substacks []*pb.SubstackProto, err error) []*pb.SubstackProto {
+	return append(substacks, &pb.SubstackProto{
+		Lines: []*pb.LineProto{
+			{
+				Line: &pb.LineProto_Error{
+					Error: &pb.ErrorLineProto{
+						Error: err.Error(),
+					},
+				},
+			},
+		},
+	})
+}
+
+func fillSubstacksFromFeature(substacks []*pb.SubstackProto, f b6.Feature, w b6.World) []*pb.SubstackProto {
+	substack := &pb.SubstackProto{}
+	substack.Lines = append(substack.Lines, valueLineFromValue(f))
+	substack.Lines = append(substack.Lines, lineFromTags(f))
+	substacks = append(substacks, substack)
+	if path, ok := f.(b6.PathFeature); ok {
+		substack := &pb.SubstackProto{Collapsable: true}
+		line := valuePairLineFromValues("Points", path.Len())
+		substack.Lines = append(substack.Lines, line)
+		for i := 0; i < path.Len(); i++ {
+			if point := path.Feature(i); point != nil {
+				substack.Lines = append(substack.Lines, valueLineFromValue(point))
+			} else {
+				substack.Lines = append(substack.Lines, valueLineFromValue(path.Point(i)))
+			}
+		}
+		substacks = append(substacks, substack)
+	}
+	relations := b6.AllRelations(w.FindRelationsByFeature(f.FeatureID()))
+	if len(relations) > 0 {
+		substack := &pb.SubstackProto{Collapsable: true}
+		line := valuePairLineFromValues("Relations", len(relations))
+		substack.Lines = append(substack.Lines, line)
+		for _, r := range relations {
+			substack.Lines = append(substack.Lines, valueLineFromValue(r))
+		}
+		substacks = append(substacks, substack)
+	}
+	return substacks
+}
+
+func fillResponseFromResult(response *UIResponseJSON, result interface{}, rules renderer.RenderRules, w b6.World) error {
+	p := (*pb.UIResponseProto)(response.Proto)
+	switch r := result.(type) {
+	case string:
+		p.Stack.Substacks = fillSubstacksFromString(p.Stack.Substacks, r)
+	case b6.Feature:
+		p.Stack.Substacks = fillSubstacksFromFeature(p.Stack.Substacks, r, w)
+		highlightInResponse(p, r.FeatureID())
+	case b6.Query:
+		if q, ok := api.UnparseQuery(r); ok {
+			p.Stack.Substacks = fillSubstacksFromString(p.Stack.Substacks, q)
+		} else {
+			// TODO: Improve the rendering of queries
+			p.Stack.Substacks = fillSubstacksFromString(p.Stack.Substacks, "Query")
+		}
+	case b6.Tag:
+		p.Stack.Substacks = fillSubstacksFromAtom(p.Stack.Substacks, atomFromValue(r))
+		if !rules.IsRendered(r) {
+			if q, ok := api.UnparseQuery(b6.Tagged(r)); ok {
+				p.QueryLayers = append(p.QueryLayers, q)
+			}
+		}
+	case *api.HistogramCollection:
+		substack := &pb.SubstackProto{}
+		if err := fillSubstackFromHistogram(substack, r); err == nil {
+			p.Stack.Substacks = append(p.Stack.Substacks, substack)
+		} else {
+			p.Stack.Substacks = fillSubstacksFromError(p.Stack.Substacks, err)
+		}
+	case api.Collection:
+		substack := &pb.SubstackProto{}
+		if err := fillSubstackFromCollection(substack, r, p); err == nil {
+			p.Stack.Substacks = append(p.Stack.Substacks, substack)
+		} else {
+			return err
```

Comment:
u dont wanna fill from error here too ?

Reply:
Good spot - actually, I check the error in the caller and fill there, so it's the case above that's incorrect, since it should be returning the error, not filling and returning nil. I've changed that.